src/frontend/core/executor/impl/TriangleCountExecutor.cpp

```diff
@@ -34,13 +50,400 @@ TriangleCountExecutor::TriangleCountExecutor(SQLiteDBInterface *db, PerformanceS
+                need_pushing = false;
+                break;
+            }
+            if (alloc.find(p) == alloc.end()) {
+                continue;
+            }
+            auto w = alloc[p];
+            if (std::find(ws.begin(), ws.end(), w) != ws.end()) {
+                alloc.erase(p);
+                alloc[p0] = w;
+                remain_l.push_back(p);
+                need_pushing = false;
+                break;
+            }
+        }
+        if (need_pushing) copying.push_back(p0);
+    }
+    return copying;
+}
+
+static void scale_up(std::map<string, int> &loads, map<string, string> &workers, int copy_cnt) {
+    int curr_load = 0;
+    for (auto it = loads.begin(); it != loads.end(); it++) {
+        curr_load += it->second;
+    }
+    int n_cores = copy_cnt + curr_load - 3 * loads.size();
+    if (n_cores < 0) {
+        return;
+    }
+    int n_workers = n_cores / 2 + 1;  // allocate a little more to prevent saturation
+    if (n_cores % 2 > 0) n_workers++;
+    if (n_workers == 0) return;
+
+    K8sWorkerController *k8sController = K8sWorkerController::getInstance();
+    map<string, string> w_new = k8sController->scaleUp(n_workers);
+
+    for (auto it = w_new.begin(); it != w_new.end(); it++) {
+        loads[it->first] = 0.1;
+        workers[it->first] = it->second;
+    }
+}
+
+static int alloc_net_plan(std::map<int, string> &alloc, std::vector<int> &parts,
+                          std::map<int, std::pair<string, string>> &transfer, std::map<string, int> &net_loads,
+                          std::map<string, int> &loads, const std::map<int, std::vector<string>> &p_avail,
+                          int curr_best) {
+    int curr_load = std::max_element(net_loads.begin(), net_loads.end(),
+                                     [](const std::map<string, int>::value_type &p1,
+                                        const std::map<string, int>::value_type &p2) { return p1.second < p2.second; })
+                        ->second;
+    if (curr_load >= curr_best) {
+        return curr_load;
+    }
+    if (parts.empty()) {
+        if (net_loads.empty()) return 0;
+        return curr_load;
+    }
+    struct best_net_alloc {
+        std::map<int, string> alloc;
+        std::map<int, std::pair<string, string>> transfer;
+        std::map<string, int> net_loads;
+        std::map<string, int> loads;
+    };
+    int best = curr_best;
+    struct best_net_alloc best_plan = {.transfer = transfer, .net_loads = net_loads, .loads = loads};
+    int p = parts.back();
+    parts.pop_back();
+    vector<string> wts;
+    for (auto it = loads.begin(); it != loads.end(); it++) {
+        wts.push_back(it->first);
+    }
+    sort(wts.begin(), wts.end(), [&loads](string &w1, string &w2) {
+        int l1 = loads[w1];
+        int l2 = loads[w2];
+        // TODO: temporarily commented for scale up only
+        // if (l1 < 3 && l2 < 3) return LOAD_PREFERENCE[l1] > LOAD_PREFERENCE[loads[w2]];
```

Comment:
Have we implemented scaling down logic as well?

Reply:
Not yet. This "scale up only" means it is not used for cases where system scales up. (There might be situations where the system can handle the load without scaling up)